tools/make_patch.c

```diff
@@ -0,0 +1,475 @@
+		}
+	} else {
+		// Parse constant's value
+		*bank = 0;
+
+		char *endptr;
+		*address = strtol(buffer, &endptr, 16);
+		if (*address < 0) {
+			*address += 0x100;
+		}
+
+		if (endptr != buffer + strlen(buffer)) {
+			error_exit("Error: Cannot parse value: %s\n", buffer);
+		}
+	}
+
+	free(buffer);
+}
+
+struct Symbol *parse_symbols(const char *filename, struct Symbol **symbols) {
+	FILE *file = xfopen(filename, 'r');
+	struct Buffer *buffer = create_buffer(1);
+
+	int bank = 0;
+	int address = 0;
+	bool skip_line = false;
+
+	for (int c = getc(file); c != EOF; c = getc(file)) {
+		switch (c) {
+		case ';':
+		case '\r':
+		case '\n':
+			// This is the end of the symbol name
+			append_to_buffer(buffer, "");
+			append_symbol(symbols, buffer->data, bank, address);
+			// Clear the buffer and skip to the next line for another symbol
+			buffer->size = 0;
+			skip_line = true;
+			break;
+
+		case ' ':
+			// This is the end of the symbol value
+			append_to_buffer(buffer, "");
+			parse_symbol_value(buffer->data, &bank, &address);
+			// Clear the buffer for the symbol name
+			buffer->size = 0;
+			break;
+
+		default:
+			append_to_buffer(buffer, &c);
+		}
+
+		if (skip_line) {
+			// Skip to the next line
+			while (c != '\n' && c != '\r' && c != EOF) {
+				c = getc(file);
+			}
+			skip_line = false;
+		}
+	}
+
+	fclose(file);
+	free_buffer(buffer);
+	return *symbols;
+}
+
+int parse_arg_value(char *arg, bool absolute, const struct Symbol *symbols, const char *patch_name) {
+	static const char *comparisons[6] = {"==", ">", "<", ">=", "<=", "!="};
+	for (unsigned int i = 0; i < sizeof(comparisons) / sizeof(*comparisons); i++) {
+		if (!strcmp(arg, comparisons[i])) {
+			return i;
+		}
+	}
+	if (isdigit((unsigned char)arg[0])) {
+		return strtol(arg, NULL, 0);
+	}
+	int offset_mod = 0;
+	const char *plus = strchr(arg, '+');
+	if (plus) {
+		offset_mod = strtol(plus, NULL, 0);
+		arg[strlen(arg) - strlen(plus)] = '\0';
+	}
+	const char *sym_name = !strcmp(arg, "@") ? patch_name : arg;
+	const struct Symbol *symbol = find_symbol(symbols, sym_name);
+	return (absolute ? symbol->offset : symbol->address) + offset_mod;
+}
+
+void interpret_command(
+	char *command, const struct Symbol *current_hook, const struct Symbol *symbols,
+	struct Buffer *patches, FILE *new_rom, FILE *orig_rom, FILE *output
+) {
+	// Strip all leading spaces and all but one trailing space
+	int x = 0;
+	for (int i = 0; command[i]; i++) {
+		if (!isspace(command[i]) || (i > 0 && !isspace(command[i-1]))) {
+			command[x++] = command[i];
+		}
+	}
+	if (isspace(command[x-1])) {
+		command[x-1] = '\0';
+	} else {
+		command[x] = '\0';
+	}
+
+	// Count the arguments
+	int argc = 0;
+	for (char *c = command; *c; c++) {
+		if (*c == ' ') {
+			argc++;
+		}
+	}
+
+	// Get the arguments
+	char *argv[argc]; // VLA
+	char *arg = command;
+	for (int i = 0; i < argc; i++) {
+		arg = strchr(arg, ' ');
+		if (!arg) {
+			break;
+		}
+		*arg++ = '\0';
+		argv[i] = arg;
+	}
```

Comment:
You might want to replace the above two loops with calls to `strsep`, which would also let you handle tokens separated by more than one instance of more than one character.

Reply:
Well, `strsep` is not portable, but I only suggested it as an alternative to `strtok`. (Wasn't there a standardized version that doesn't have that weird `NULL` API?)